packages/core/src/utilities/isEqual.ts

```diff
@@ -1,27 +1,72 @@
+function areNumbersEqualWithTolerance(
+  num1: number,
+  num2: number,
+  tolerance: number
+): boolean {
+  return Math.abs(num1 - num2) <= tolerance;
+}
+
+function areArraysEqual(
+  arr1: number[] | Float32Array | Float64Array,
+  arr2: number[] | Float32Array | Float64Array,
+  tolerance = 1e-5
+): boolean {
+  if (arr1.length !== arr2.length) {
+    return false;
+  }
+
+  for (let i = 0; i < arr1.length; i++) {
+    if (!areNumbersEqualWithTolerance(arr1[i], arr2[i], tolerance)) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+function isNumberType(value: any): value is number {
+  return typeof value === 'number';
+}
+
+function isFloatArrayLike(
+  value: any
+): value is number[] | Float32Array | Float64Array {
+  const isArrayLike = typeof value === 'object' || 'length' in value;
+  const floatTypedArrayLike =
+    value instanceof Float32Array || value instanceof Float64Array;
+
+  return isArrayLike || floatTypedArrayLike;
+}
+
 /**
- * returns equal if the two arrays are identical within the
- * given tolerance.
+ * Returns whether two values are equal or not based on epsilon comparison.
+ * For array comparison, it does NOT strictly compare them but only compare its values.
+ * Typed Array comparison is valid for Float32Array and Float16Array only.
  *
- * @param v1 - The first array of values
- * @param v2 - The second array of values.
+ * @param v1 - The first value to compare
+ * @param v2 - The second value to compare
  * @param tolerance - The acceptable tolerance, the default is 0.00001
  *
  * @returns True if the two values are within the tolerance levels.
  */
-export default function isEqual(
-  v1: number[] | Float32Array,
-  v2: number[] | Float32Array,
+export default function isEqual<ValueType>(
+  v1: ValueType,
+  v2: ValueType,
   tolerance = 1e-5
 ): boolean {
-  if (v1.length !== v2.length) {
+  // values must be the same type
+  if (typeof v1 !== typeof v2) {
     return false;
   }
 
-  for (let i = 0; i < v1.length; i++) {
-    if (Math.abs(v1[i] - v2[i]) > tolerance) {
-      return false;
-    }
+  if (isNumberType(v1) && isNumberType(v2)) {
```

Comment:
I guess we only need to check one of them, since you return false early in the function right? same for below

Reply:
Remember when we were seeing support for es stuff? 
There there was the concept of narrowing.
https://www.typescriptlang.org/docs/handbook/2/narrowing.html

To answer your question narrowing op is not transitive in TS yet. So we need to narrow both params to ensure the if block has the type we want